packages/devtools_testing/lib/riverpod/controller_test.dart

```diff
@@ -0,0 +1,190 @@
+// Copyright 2021 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// ignore_for_file: implementation_imports, invalid_use_of_visible_for_testing_member, non_constant_identifier_names
+
+import 'package:devtools_app/src/instance_viewer/instance_providers.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:devtools_app/src/eval_on_dart_library.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:devtools_app/src/globals.dart';
+import 'package:devtools_app/src/riverpod/provider_list.dart';
+import 'package:meta/meta.dart';
+import 'package:vm_service/vm_service.dart';
+
+import '../support/flutter_test_environment.dart';
+
+typedef CancelSubscriptionCallback = Future<void> Function();
+
+Future<void> runRiverpodControllerTests(FlutterTestEnvironment env) async {
+  group('runRiverpodControllerTests', () {
+    EvalOnDartLibrary eval;
+    IsAlive isAlive;
+
+    setUp(() async {
+      await env.setupEnvironment();
+      await serviceManager.service.allFuturesCompleted;
+
+      isAlive = IsAlive();
+      eval = EvalOnDartLibrary(
+        [
+          'package:riverpod_app/main.dart',
+          'package:riverpod_app/tester.dart',
+        ],
+        env.service,
+      );
+    });
+
+    Future<CancelSubscriptionCallback> listenToProvider({
+      @required InstanceRef containerRef,
+      @required String provider,
+      Map<String, String> scope,
+    }) async {
+      final subscriptionRef = await eval.safeEval(
+        'container.listen($provider)',
+        isAlive: isAlive,
+        scope: {...scope, 'container': containerRef.id},
+      );
+
+      await eval.awaitEval('tester.pump()', isAlive: isAlive);
```

Comment:
add a comment explaining why you need to call tester.pump on the app being tested.